docs/refman/ppr-followme.1.pod
#
# mouse:~ppr/src/docs/refman/ppr-followme.1.pod
# Copyright 1995--2003, Trinity College Computing Center.
# Written by David Chappell.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:
#
# * Redistributions of source code must retain the above copyright notice,
# this list of conditions and the following disclaimer.
# 
# * Redistributions in binary form must reproduce the above copyright
# notice, this list of conditions and the following disclaimer in the
# documentation and/or other materials provided with the distribution.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS OR CONTRIBUTORS BE 
# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR 
# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF 
# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS 
# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN 
# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) 
# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE 
# POSSIBILITY OF SUCH DAMAGE.
#
# Last modified 10 March 2003.
#

=head1 NAME

B<ppr-followme> - Inform responder system of your current address

=head1 SYNOPSIS

B<ppr-followme> [B<--version>] [B<--help>] [I<responder> [I<address>]]

=head1 DESCRIPTION



=head1 FILES

This program creates files in F</var/spool/ppr/followme.db/>.

=head1 SEE ALSO

"PPR, a PostScript Print Spooler"

=head1 HISTORY

PPR was written at Trinity College during 1993--2003.

=head1 AUTHOR

David Chappell, Trinity College Computing Center, Hartford, Connecticut.

=cut


History.txt
mouse:~ppr/src/History.txt (symbolic link)

This file contains a history of changes to PPR which occured after the
release of version 1.00.


1.00, 26 April 1995

First public release.


1.10, 27 April 1995--7 June 1995

Removed "From:" line from message passed to Sendmail by pprd when sending
a message listing the remaining printer faults.  This is so that Sendmail
will take the "real name" from the password file and put it in the "From:"
line.

Removed dotmatrix/input.epson, a GIF file which had been inadvertently
included in the distribution.

Moved the contents of include/atalk_streams.h into
interfaces/atalk_streams.c since it was not long enough to justify an
include file, nor was it of interest to any other program.

Added code to ppad to classify members of the group "ppradmin" as PPR
addministrators who can edit printer and group configurations.

Added -v switch to ppad.

Removed uid's 1-99 from the list of priveledged users.

Fixed ppad(8) so that it considers the group "ppradmin" as priveledged
rather than "ppr".

Changed ppr(1) so that it will accept the members of the group "pprprox" as
priveledged as well as the members of "ppr".  This is for security reasons
as one may not want to grant a user who has the privedge of using the -f
switch the priveledge of deleting other people's jobs.

Modified ppop(1) and pprd(8) so that user's can request to see the queue
entry for a single job.  Previously, the only options were all jobs for a
certain destination or all jobs in all queues.

Rewrote and renamed the undocumented "ppop qdebug" command to become the new
documented command "ppop details".  (This was done primarily to support
wwwppop, an experimental managment interface for PPR.  Since it was still
incomplete it was not released with this version of PPR.)

Renamed interfaces/atalk_streams.c to atalk_ali.c and
papsrv/papsrv_streams.c to papsrv_ali.c

Removed code from "make dirs" so that it doesn't create the directory
"/etc/ppr/groups.ppd".

Added sleep=N and eoftimeout=N options to tcpip interface.

Added -C _filename_ switch to ppr invokation in xppr.

Added "-default okay" to xmessage invokation in xppr.

Added the "ppop lpq" sub-command to the "ppop help" information, thus
making it a documented sub-command.

Extended automatic bin selection to multiple media types when "%%PageMedia:"
comments are present.  This has not yet been well tested.

Made "%%EndProcset" an acceptable substitute for "%%EndProcSet" due to a
reputed bug in QuickDraw GX.

Changed rule in ppr_infile.c for detecting PCL code so that documents with
only one escape would still be detected.


1.20, 8 June 1995--10 January 1996

Fixed code in ppr_infile.c so that the ppr -o switch will work even if there
is no "DefFiltOpts:" line in the printer or group configuration file.

Added '(' to list of likely second characters for HP escape codes
in ppr_infile.c

Replaced printing of errno value with strerror(errno) output in some places,
eliminated printing of errno=x from other places where both were printed.

Fixed ppr(1) so that it sets the environment variables "PATH", "IFS", and
"SHELL" so that filters and responders it invokes are less easy to subvert.

Removed setting of PATH from ~ppr/src/misc_filters/tex.perl and changed
shell name from "/usr/bin/tainperl" to "/usr/bin/perl".

Changed code which determins default responder address so that the
environment variables "USER" and "LOGNAME" are used in preference to the
current uid.

Added prompts in Configure to set the directory into which the links to PPR
binaries will be written and the path to the X-Windows binaries such as
xterm.

Modified papsrv so that it will pass the network and node number of the
client to ppr with the -r switch.  (This is for use with an experimental
AppleTalk responder.)

Modified papsrv.c, renaming cleanup() to termination_handler() and causing
it to call fatal() only the first time it is called.  Without this, name
could be left on the network if the CAP version received two terminate
signals in rapid sucession.

Added code to re-line-wrap responder messages for certain responders.
(The library function responder_width() returns the width for a responder.
Currently it is hard coded to return zero for all responders except "samba".
A return value of zero indicates that the message should not be
re-line-wrapped.)

Dropt reading of LPDEST and PRINTER by ppr(1).  (I can not really remember
why I put this feature in.  I don't think the reasons where very good.  It
may have been that I thought that the same printer would be available thru
two spooling systems so one might want to set the default destination for
both with one variable.  That now seems a little far fetched.)

Separated Solaris and Starserver sections of sysdep.h

Added PNG (Portable Network Graphics) format with magic number recognition
to ppr_infile.c.  This remains untested since I do not have any programs
which manipulate this graphics format.

In pprd.c, put sleep(1) between execl("pprdrv") and _exit(242) so that pprd
has time to record the PID before the child exits.  (If this is not done
and execl() fails immediately, a child which should have become pprdrv may
exit before the parent records its PID in the information structure for that
queue.  This result is that it will not recoginize the purpose of the child
when it receives notification of its death and will leave the queue in the
printing state indefinitely.)

Did same in pprdrv.c in code which executes interface.  Also changed
_exit(255) to _exit(242).  (Exit 242 is the new standard since PERL uses 255
as the exit code when die() is executed.)

Added automatic creation of "DefFiltOpts:" lines to ppad(8).  Manual
creation is supported by the the "ppad deffiltopts" and "ppad group
deffiltopts" command.

Modified ppop so that interactive mode restores the terminal settings when
it exits.  Also changed terminal setting code to use the POSIX method of

unixuser/ppr-followme.c
/*
** mouse:~ppr/src/unixuser/ppr-followme.c
** Copyright 1995--2003, Trinity College Computing Center.
** Written by David Chappell.
**
** Redistribution and use in source and binary forms, with or without
** modification, are permitted provided that the following conditions are met:
**
** * Redistributions of source code must retain the above copyright notice,
** this list of conditions and the following disclaimer.
**
** * Redistributions in binary form must reproduce the above copyright
** notice, this list of conditions and the following disclaimer in the
** documentation and/or other materials provided with the distribution.
**
** THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
** AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
** IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
** ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDERS OR CONTRIBUTORS BE
** LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
** CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
** SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
** INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
** CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
** ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
** POSSIBILITY OF SUCH DAMAGE.
**
** Last modified 10 October 2003.
*/

#include "before_system.h"
#include <stdio.h>
#include <stdlib.h>
#include <pwd.h>
#include <unistd.h>
#include <errno.h>
#include <string.h>
#ifdef INTERNATIONAL
#include <locale.h>
#include <libintl.h>
#endif
#include "gu.h"
#include "global_defines.h"
#include "util_exits.h"
#include "version.h"

const char myname[] = "ppr-followme";

const char dbdir[] = VAR_SPOOL_PPR"/followme.db";

/*
** This function creates the record which is later picked up by ppr-respond.
*/
static int write_record(const char username[], const char responder[], const char responder_address[], const char responder_options[])
	{
	FILE *f;
	char fname[MAX_PPR_PATH];

	ppr_fnamef(fname, "%s/%s", dbdir, username);
	if(!(f = fopen(fname, "w")))
		{
		fprintf(stderr, _("%s: fopen(\"%s\", \"w\") failed, errno=%d (%s)\n"), myname, fname, errno, gu_strerror(errno));
		return -1;
		}

	fprintf(f, "%s %s \"%s\"\n", responder, responder_address, responder_options);

	fclose(f);

	return 0;
	}

/*
** This function reads one's existing followme record and prints it.
*/
static int do_show(const char username[])
	{
	FILE *f;

	{
	char fname[MAX_PPR_PATH];
	ppr_fnamef(fname, "%s/%s", dbdir, username);
	if(!(f = fopen(fname, "r")))
		{
		if(errno == ENOENT)
			{
			printf(_("The user \"%s\" is not registered with followme.\n"), username);
			return EXIT_NOTFOUND;
			}
		else
			{
			fprintf(stderr, _("%s: fopen(\"%s\", \"r\") failed, errno=%d (%s)\n"), myname, fname, errno, gu_strerror(errno));
			return EXIT_INTERNAL;
			}
		}
	}

	{
	char *responder = NULL, *responder_address = NULL, *responder_options = NULL;
	{
	char *line = NULL;
	int line_space_available = 80;

	if(!(line = gu_getline(line, &line_space_available, f)) || gu_sscanf(line, "%S %S %Q", &responder, &responder_address, &responder_options) != 3)
		{
		fprintf(stderr, _("%s: invalid followme registration record\n"), myname);
		return EXIT_INTERNAL;
		}

	gu_free(line);
	}

	printf("%s %s \"%s\"\n", responder, responder_address, responder_options);
	gu_free(responder);
	gu_free(responder_address);
	gu_free(responder_options);
	}

	return EXIT_OK;
	}

/*
** This takes up to three command line arguments and writes them to a
** followme record.
*/
static int do_set(char username[], int argc, char *argv[], int i)
	{
	char *responder = NULL, *responder_address = NULL, *responder_options = NULL;

	/* Read up to three command line options. */
	if((argc - i) >= 1)
		responder = argv[i + 0];
	if((argc - i) >= 2)
		responder_address = argv[i + 1];
	if((argc - i) >= 3)
		responder_options = argv[i + 2];

	/* Now we supply defaults for any that weren't specified. */

	if(!responder)
		{
		if(getenv("DISPLAY"))
			responder = "xwin";
		else
			responder = "write";
		}

	if(!responder_address)
		{
		if(strcmp(responder, "xwin") == 0)
			{
			if(!(responder_address = getenv("DISPLAY")))
				responder_address = ":0.0";
			}
		else if(strcmp(responder, "pprpopup") == 0)
			gu_asprintf(&responder_address, "%s@localhost", username);
		else
			responder_address = username;
		}

	if(!responder_options)
		if(!(responder_options = getenv("PPR_RESPONDER_OPTIONS")))
			responder_options = "";

	/* Detect settings that could subvert the user's intentions. */
	{
	char *p;
	if((p = getenv("PPR_RESPONDER")) && strcmp(p, "followme") != 0)
		fprintf(stderr, _("Warning: PPR_RESPONDER is set to %s\n"), p);
	if((p = getenv("PPR_RESPONDER_ADDRESS")) && strcmp(p, username) != 0)
		fprintf(stderr, _("Warning: PPR_RESPONDER_ADDRESS is set to %s\n"), p);
	}

	/* Show the user what we have decided on. */
	printf("%s %s \"%s\"\n", responder, responder_address, responder_options);

	/* Save it in followme.db. */
	if(write_record(username, responder, responder_address, responder_options) == -1)
		return EXIT_INTERNAL;

	/* The xwin responder won't work unless we give it X display access. */
	if(strcmp(responder, "xwin") == 0)
		{
		gu_runl(myname, stderr, HOMEDIR"/bin/ppr-xgrant", NULL);
		}

	return EXIT_OK;
	} /* end of do_set() */

/*
** Command line options:
*/
static const char *option_chars = "";
static const struct gu_getopt_opt option_words[] =
		{
		{"show", 1000, FALSE},
		{"help", 9000, FALSE},
		{"version", 9001, FALSE},
		{(char*)NULL, 0, FALSE}
		} ;

/*
** Print help.
*/
static void help_usage(FILE *outfile)
	{
	fprintf(outfile, _("Usage: %s [switches]\n"), myname);

	fprintf(outfile, _("       %s [responder [address [\"options\"]]]\n"), myname);

	fputc('\n', outfile);

	fputs(_("Valid switches:\n"), outfile);

	fputs(_("\t--show\n"), outfile);

	fputs(_("\t--version\n"
						"\t--help\n"), outfile);
	}

int main(int argc, char *argv[])
	{
	int i;
	struct passwd *pw;
	gu_boolean opt_show = FALSE;

	/* Initialize international messages library. */
	#ifdef INTERNATIONAL
	setlocale(LC_ALL, "");
	bindtextdomain(PACKAGE, LOCALEDIR);
	textdomain(PACKAGE);
	#endif

	/* Parse the options. */
	{
	struct gu_getopt_state getopt_state;
	int optchar;
	gu_getopt_init(&getopt_state, argc, argv, option_chars, option_words);
	while((optchar = ppr_getopt(&getopt_state)) != -1)
		{
		switch(optchar)
			{
			case 1000:					/* --show */
				opt_show = TRUE;
				break;

			case 9000:					/* --help */
				help_usage(stdout);
				return EXIT_OK;

			case 9001:					/* --version */
				puts(VERSION);
				puts(COPYRIGHT);
				puts(AUTHOR);
				return EXIT_OK;

			default:					/* other getopt errors or missing case */
				gu_getopt_default(myname, optchar, &getopt_state, stderr);
				return EXIT_SYNTAX;
			}
		}
	i = getopt_state.optind;
	}

	/* We have no use for more than 4 non-switch arguments. */
	if((argc - i) >= 4)
		{
		help_usage(stderr);
		return EXIT_SYNTAX;
		}

	/* We need the username since it is the key into followme.db. */
	if(!(pw = getpwuid(getuid())))
		{
		fprintf(stderr, X_("%s: getpwuid(%ld) failed, you are a non-user\n"), myname, (long)getuid());
		return EXIT_INTERNAL;
		}

	if(opt_show)
		return do_show(pw->pw_name);
	else
		return do_set(pw->pw_name, argc, argv, i);

	} /* end of main() */

/* end of file */
